File: leetCode.com/typescript/strings/implementStrStr().ts
// Given two strings needle and haystack, return the index of the first occurrence of needle in haystack, or -1 if needle is not part of haystack.

function strStr1(haystack: string, needle: string): number {
    return haystack.indexOf(needle)
};

// Manual
function strStr2(haystack: string, needle: string): number {
    const n = haystack.length;
    const m = needle.length;

    if (m === 0) {
        return 0;
    }

    for (let i = 0; i <= n - m; i++) {
        if (haystack.substring(i, i + m) === needle) {
            return i;
        }
    }
    return -1;
}

File: leetCode.com/python/strings/implementStrStr().py
# Given two strings needle and haystack, return the index of the first occurrence of needle in haystack, or -1 if needle is not part of haystack.

def strStr1(haystack: str, needle: str) -> int:
    return haystack.find(needle)

# Manual
def strStr2(haystack: str, needle: str) -> int:
    n, m = len(haystack), len(needle)
    if m == 0:
        return 0

    for i in range(n - m + 1):
        if haystack[i:i+m] == needle:
            return i
    return -1
